working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1748,65 +1746,196 @@ types of all of the case expressions.
+    block or function body, but it is a compile-time error to refer to them
+    until after the pattern variable declaration's initializer:*
+
+    ```dart
+    const c = 1;
+
+    f() {
+      print(c);
+      //    ^ Error: Refers to C declared below:
+
+      var [c] = c;
+      //        ^ Error: Not initialized yet.
+
+      print(c);
+      //    ^ OK.
+    }
+    ```
+
+*   **Pattern-for statement**: Scoping follows the normal for and for-in
+    statement scoping rules where the variable (now variables) are bound in a
+    new scope for each loop iteration. All pattern variables are in the same
+    scope. They are considered initialized after the for loop initializer
+    expression.
+
+*   **Pattern assignment**: An assignment only assigns to existing variables
+    and does not bind any new ones.
+
+*   **Switch statement**, **switch expression**, **if-case statement**: Each
+    `caseHead` introduces a new *case scope* which is where the variables
+    defined by that case's pattern are bound.
+
+    There is no *initializing expression* for the variables in a case pattern,
+    but they are considered initialized after the entire case pattern, before
+    the guard expression if there is one. *However, all pattern variables are
+    in scope in the entire pattern:*
+
+    ```dart
+    const c = 1;
+    switch (1) {
+      case [var c, == c]
+        //            ^ Error: In scope but not initialized.
+        //              (Also an error because `c` is not a constant.)
+            when c == 2:
+        //       ^ OK.
+        print(c);
+        //    ^ OK.
+    }
+    ```
 
-The scope where a pattern's variables are declared and available for use depends
-on the construct that contains the pattern:
+    The guard expression is evaluated in its case's case scope.
 
-*   **Local pattern variable declaration**: The rest of the block following
-    the declaration.
-*   **For loop pattern variable declaration**: The body of the loop and the
-    condition and increment clauses in a C-style for loop.
-*   **Switch statement case**: The guard clause and the statements of the
-    subsequent non-empty case body.
-*   **Switch expression case**: The guard clause and the case expression.
-*   **If-case statement**: The guard clause and then statement.
+#### Switch case fallthrough
 
-Multiple switch case patterns may share the same variable scope if their case
-bodies are empty:
+In a switch statement or expression, multiple cases may share the same body:
 
 ```dart
 switch (obj) {
   case [int a, int b]:
   case {"a": int a, "b": int b}:
     print(a + b); // OK.
+}
+```
 
-  case [int a]:
-  case (String a): // Error.
-    break;
+This would normally be a name collision, but we make an exception to allow it
+when it doesn't cause any problems.
+
+When compiling the body shared by a set of cases with pattern variable sets `vs`
+(where a default case has an empty pattern variable set), the enclosing scope is
```

Comment:
Should also mention that a label has an empty pattern variable set.

Reply:
Ugh labels. Good catch.